File: packages/runtime-web-three/src/countdowns.ts
import type { IIrCountdownDeclaration, ISystemsIr, IWorldIr } from "@threenative/ir";

export interface ICountdownObservation {
  countdown: string;
  event: string;
  fired: boolean;
  tick: number;
  value: number;
}

interface ICountdownRuntimeEntry {
  fired: boolean;
  initialized: boolean;
  restartToken: unknown;
  running: boolean;
}

export interface ICountdownRuntimeState {
  entries: Map<string, ICountdownRuntimeEntry>;
}

export function createCountdownRuntimeState(): ICountdownRuntimeState {
  return { entries: new Map() };
}

export function stepCountdowns(
  world: IWorldIr,
  systems: ISystemsIr,
  fixedDelta: number,
  runtime: ICountdownRuntimeState,
  tick: number,
): ICountdownObservation[] {
  const delta = Number.isFinite(fixedDelta) && fixedDelta > 0 ? fixedDelta : 0;
  const observations: ICountdownObservation[] = [];
  for (const countdown of systems.countdowns ?? []) {
    const resource = world.resources?.[countdown.resource];
    if (!isRecord(resource)) {
      continue;
    }
    const entry = runtime.entries.get(countdown.id) ?? {
      fired: false,
      initialized: false,
      restartToken: undefined,
      running: countdown.autostart !== false,
    };
    const restartToken = resource.restartToken;
    const restart = entry.initialized && !sameValue(entry.restartToken, restartToken);
    const running = typeof resource.running === "boolean" ? resource.running : (countdown.autostart !== false);
    if (restart || (!entry.running && running)) {
      resource[countdown.field] = startValue(countdown);
      entry.fired = false;
    }
    entry.initialized = true;
    entry.restartToken = restartToken;
    entry.running = running;
    if (!running) {
      runtime.entries.set(countdown.id, entry);
      continue;
    }
    const current = finite(resource[countdown.field], startValue(countdown));
    const value = countdown.direction === "down"
      ? Math.max(0, Math.min(countdown.limit, current - delta))
      : Math.min(countdown.limit, Math.max(0, current + delta));
    resource[countdown.field] = round(value);
    const reached = countdown.direction === "down" ? value <= 0 : value >= countdown.limit;
    const fired = reached && !entry.fired;
    if (fired) {
      entry.fired = true;
      appendEvent(world, countdown.event, {
        countdown: countdown.id,
        direction: countdown.direction,
        field: countdown.field,
        limit: countdown.limit,
        resource: countdown.resource,
        value: resource[countdown.field],
      });
    }
    runtime.entries.set(countdown.id, entry);
    observations.push({ countdown: countdown.id, event: countdown.event, fired, tick, value: resource[countdown.field] as number });
  }
  return observations;
}

export function resetCountdowns(runtime: ICountdownRuntimeState): void {
  runtime.entries.clear();
}

function startValue(countdown: IIrCountdownDeclaration): number {
  return countdown.direction === "down" ? countdown.limit : 0;
}

function appendEvent(world: IWorldIr, event: string, payload: Record<string, unknown>): void {
  const queue = world.events?.[event];
  world.events = { ...(world.events ?? {}), [event]: Array.isArray(queue) ? [...queue, payload] : [payload] };
}

function finite(value: unknown, fallback: number): number {
  return typeof value === "number" && Number.isFinite(value) ? value : fallback;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function sameValue(left: unknown, right: unknown): boolean {
  return JSON.stringify(left) === JSON.stringify(right);
}

function round(value: number): number {
  return Number(value.toFixed(6));
}

File: runtime-bevy/crates/threenative_runtime/src/countdowns.rs
use std::collections::BTreeMap;

use serde_json::{Value, json};
use threenative_loader::LoadedBundle;

#[derive(Clone, Debug, Default, PartialEq)]
pub struct NativeCountdownRuntimeState {
    entries: BTreeMap<String, NativeCountdownRuntimeEntry>,
}

#[derive(Clone, Debug, PartialEq)]
struct NativeCountdownRuntimeEntry {
    fired: bool,
    initialized: bool,
    restart_token: Option<Value>,
    running: bool,
}

#[derive(Clone, Debug, PartialEq)]
pub struct NativeCountdownObservation {
    pub countdown: String,
    pub event: String,
    pub fired: bool,
    pub tick: u64,
    pub value: f32,
}

pub fn step_bundle_countdowns(
    bundle: &mut LoadedBundle,
    fixed_delta: f32,
    tick: u64,
    runtime: &mut NativeCountdownRuntimeState,
) -> Vec<NativeCountdownObservation> {
    let delta = if fixed_delta.is_finite() && fixed_delta > 0.0 {
        fixed_delta
    } else {
        0.0
    };
    let declarations = bundle
        .systems
        .as_ref()
        .map(|systems| systems.countdowns.clone())
        .unwrap_or_default();
    let mut observations = Vec::new();
    for countdown in declarations {
        let Some(resource) = bundle
            .world
            .resources
            .get_mut(&countdown.resource)
            .and_then(Value::as_object_mut)
        else {
            continue;
        };
        let entry = runtime
            .entries
            .entry(countdown.id.clone())
            .or_insert_with(|| NativeCountdownRuntimeEntry {
                fired: false,
                initialized: false,
                restart_token: None,
                running: countdown.autostart != Some(false),
            });
        let restart_token = resource.get("restartToken").cloned();
        let restart = entry.initialized && entry.restart_token != restart_token;
        let running = resource
            .get("running")
            .and_then(Value::as_bool)
            .unwrap_or(countdown.autostart != Some(false));
        if restart || (!entry.running && running) {
            resource.insert(
                countdown.field.clone(),
                json!(start_value(&countdown.direction, countdown.limit)),
            );
            entry.fired = false;
        }
        entry.initialized = true;
        entry.restart_token = restart_token;
        entry.running = running;
        if !running {
            continue;
        }
        let current = resource
            .get(&countdown.field)
            .and_then(Value::as_f64)
            .map(|value| value as f32)
            .filter(|value| value.is_finite())
            .unwrap_or_else(|| start_value(&countdown.direction, countdown.limit));
        let value = if countdown.direction == "down" {
            (current - delta).clamp(0.0, countdown.limit)
        } else {
            (current + delta).clamp(0.0, countdown.limit)
        };
        let value = round(value);
        resource.insert(countdown.field.clone(), json!(value));
        let reached = if countdown.direction == "down" {
            value <= 0.0
        } else {
            value >= countdown.limit
        };
        let fired = reached && !entry.fired;
        if fired {
            entry.fired = true;
            append_event(
                bundle,
                &countdown.event,
                json!({
                    "countdown": countdown.id,
                    "direction": countdown.direction,
                    "field": countdown.field,
                    "limit": countdown.limit,
                    "resource": countdown.resource,
                    "value": value,
                }),
            );
        }
        observations.push(NativeCountdownObservation {
            countdown: countdown.id,
            event: countdown.event,
            fired,
            tick,
            value,
        });
    }
    observations
}

pub fn reset_countdowns(runtime: &mut NativeCountdownRuntimeState) {
    runtime.entries.clear();
}

fn append_event(bundle: &mut LoadedBundle, event: &str, payload: Value) {
    let queue = bundle
        .world
        .events
        .entry(event.to_owned())
        .or_insert_with(|| Value::Array(Vec::new()));
    if let Value::Array(values) = queue {
        values.push(payload);
    } else {
        *queue = Value::Array(vec![payload]);
    }
}

fn start_value(direction: &str, limit: f32) -> f32 {
    if direction == "down" { limit } else { 0.0 }
}

fn round(value: f32) -> f32 {
    (value * 1_000_000.0).round() / 1_000_000.0
}
